### PR TITLE
Training of GC Network -> debugging 

### LINKED ISSUE #147: [BUG]: Unable to train the model successfully
### Describe the bug

Training doesn't seem to proceed well - accuracy saturates after a few epochs. Accuracy & confusion matrix do not match. 

Training log [2023-08-22_17-34-27]:

Screenshot 2023-08-22 at 17 48 38

Similarly, the confusion matrices always look like below:

Screenshot 2023-08-23 at 11 38 36

### LINKED ISSUE #187: Include a `self.predict()` method into the GCN classifier
Run `predict` for inference, setting dropout to 0.0.

### LINKED ISSUE #190: [BUG] Fix how the subgraph dataset is created at inference time
Current code throws errors if `mode == "whole"` is selected because `sub_graph` is not defined (and shouldn't be):

https://github.com/alan-turing-institute/grace/blob/e634d3621747d975f974aade96141bb9174d5594/grace/models/datasets.py#L81-L92

### LINKED ISSUE #193: Rename `test` to `valid` function in `train.py`

### LINKED ISSUE #202: Shuffle the dataset before train-to-valid split
To avoid training / validating on data with unwanted correlations / relationships, _e.g._ on one single line.

### LINKED ISSUE #208: Allow empty `hidden_channels` list config hyperparam to create model with no GC layers
Currently, tests don't pass when GC layers are commented out / not supplied.

### LINKED ISSUE #209: Use `torch.nn.ModuleList` to include GC layers in model summary
Using in-built python lists disables printing proper summary of the model's dynamic GC layers:

```
GCN(
  (node_classifier): Linear(in_features=4, out_features=2, bias=True)
  (edge_classifier): Linear(in_features=8, out_features=2, bias=True)
)
```

Replace this by using `torch.nn.ModuleList` for this purpose to allow this:

```
GCN(
  (conv_layer_list): ModuleList(
    (0): GCNConv(2048, 512)
    (1): GCNConv(512, 64)
  )
  (node_classifier): Linear(in_features=64, out_features=2, bias=True)
  (edge_classifier): Linear(in_features=128, out_features=2, bias=True)
)
```

### PR DESCRIPTION
**Pull request by KristinaUlicna**
## PR contribution summary

**Why is this PR useful / good for?** *Please describe the problem(s) you're trying to address.*
* debugging of the GCN classifier training
* finding optimal config:

```
"keep_unknown_labels": "False",
    "patch_augs": "[]",
    "patch_aug_params": "[]",
    "patch_size": "(224, 224)",
    "keep_patch_fraction": "1.0",
    "train_to_valid_split": "0.85",
    "feature_dim": "2048",
    "num_node_classes": "2",
    "num_edge_classes": "2",
    "epochs": "500",
    "hidden_channels": "[]",
    "metrics": "['accuracy', 'confusion_matrix']",
    "dropout": "0.0",
    "batch_size": "512",
    "learning_rate": "0.05",
    "tensorboard_update_frequency": "1"
```
🟣 lr = 0.001, 🟢 lr = 0.01, 🟠 lr = 0.05:
Screenshot 2023-09-07 at 17 48 54

### List of proposed changes / linked issues & discussions

- [x] Resolves #147 
- [x] Resolves #187 
- [x] Resolves #190 
- [x] Resolves #202 
- [x] Resolves #208 
- [x] Resolves #209 
- [x] Resolves #193 

### What should a reviewer concentrate their feedback on?

* 🏃 Scripts check
* 💻 Code quality
* Replication of training results / loss + metrics

#### What type of PR is this? (check all applicable)

* 🐛 Bug fix
* #️⃣ Documentation / code annotation
* 🧑‍💻 Code refactor / style
* 🔥 Performance Improvements

#### Added tests?

* 🙅 no, because they aren't needed

---

## PR review summary

Describe what this PR does & **how you reviewed** the individual items, where needed:

* <TEXT>

*Some helper checks to tick off:*

- Focus on image annotation
- Focus on model training
- Could any optimization be applied?
- Is there any redundant code?
- Are there any spelling errors?

**In conclusion**, after my review, I'd like to:

- 🙋 ask some clarifying questions
- 🙅 suggest some specific changes

### CODE REVIEW

**Comment by crangelsmith** [comment]
It looks and runs without issues, and results are pretty good for both both edges and nodes!

 
Screenshot 2023-09-08 at 11 50 06

**Review by crangelsmith — APPROVED** [review]

**Comment by KristinaUlicna** [comment]
Forgot to add one more issue which this PR closes:
- [x] Resolves #186
